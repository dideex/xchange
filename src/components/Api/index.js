import {noty} from '../common'

export class Api {
  constructor() {
    this.backend = 'http://localhost:3030'
    this.isNetworkError = false
  }

  post(endpoint, data = {}, token = null) {
    return fetch(`${this.backend}/api/${endpoint}`, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: `bearer ${token}`,
      },
      body: JSON.stringify(data),
    }).then(response => response.json())
  }

  get(endpoint, params = '', token = null) {
    return fetch(`${this.backend}/api/${endpoint}${params}`, {
      method: 'GET',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: `bearer ${token}`,
      },
    }).then(response => response.json())
  }
  
  errorEmitter = fn => ({err, errCode, ...data}) => {
    if (err) {
      this.isNetworkError = true
      console.error('errorEmitter ___ ', errCode)
      noty(err, 'error')
    } else fn(data)
  }
}

export default new Api()
