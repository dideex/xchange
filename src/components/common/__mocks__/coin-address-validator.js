const WAValidator = () => ({
  validate: () => {},
  default: {
    validate: () => {},
  },
})

export default WAValidator
