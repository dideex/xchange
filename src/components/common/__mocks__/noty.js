class Noty {
  show() {}
}

Noty.__proto__.constructor = jest.fn()
export default Noty
