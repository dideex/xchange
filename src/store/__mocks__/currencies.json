{
  "data": [
    {
      "id": 0,
      "_id": "5bdc3a492fb62979e3105534",
      "base": "",
      "change": "0",
      "icon": "Bitcoin",
      "label": "BTC",
      "minimal": "0.1",
      "name": "Bitcoin",
      "order": 1,
      "price_usd": "4",
      "reserve": "10000",
      "source": "zzzZzzzzz",
      "type": "crypto"
    },
    {
      "id": 1,
      "_id": "5bdc3a492fb62979e3105532",
      "base": "",
      "change": "18.25",
      "icon": "Ethereum",
      "label": "ETH",
      "minimal": "1",
      "name": "Ethereum",
      "order": 2,
      "price_usd": "2",
      "reserve": "10000",
      "source": "1234 1234 1234 1234",
      "type": "crypto"
    },
    {
      "id": 2,
      "_id": "5bdc3a492fb62979e310553c",
      "base": "Ruble",
      "change": "0.00013913743704956245",
      "icon": "Sberbank",
      "label": "RUR",
      "mask": "____ ____ ____ ____",
      "minimal": "1000",
      "name": "Сбербанк руб",
      "order": 3,
      "price_usd": "1",
      "reserve": "1000000",
      "source": "4321 4321 4321 4321"
    }
  ],
  "userRate": 0.9
}
