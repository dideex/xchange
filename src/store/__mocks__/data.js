export const fakeLocale = 'fake-FK'
export const fakeToken = 'fakeToken'
export const fakeUser = {username: 'Batman', password: 'Nopassword'}
export const fakeUserData = {
  password: fakeUser.password,
  wallets: [{bitcoin: 1234}],
  lastOperations: [],
  username: fakeUser.username,
  email: fakeUser.username,
  login: fakeUser.username,
  convertedAmount: 1000,
}
