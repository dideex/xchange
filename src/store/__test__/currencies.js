export default [
  {
    id: 0,
    _id: '5bdc3a492fb62979e3105534',
    base: '',
    change: '0',
    icon: 'Bitcoin',
    label: 'BTC',
    minimal: '0.1',
    name: 'Bitcoin',
    order: 1,
    price_usd: '1000',
    reserve: '10000',
    source: 'zzzZzzzzz',
    type: 'crypto',
  },
  {
    id: 1,
    _id: '5bdc3a492fb62979e3105532',
    base: '',
    change: '18.25',
    icon: 'Ethereum',
    label: 'ETH',
    minimal: '1',
    name: 'Ethereum',
    order: 2,
    price_usd: '500',
    reserve: '10000',
    source: '1234 4321 1234 4321',
    type: 'crypto',
  },
  {
    id: 2,
    _id: '5bdc3a492fb62979e310553c',
    base: 'Ruble',
    change: '0.00013913743704956245',
    icon: 'Sberbank',
    label: 'RUR',
    mask: '____ ____ ____ ____',
    minimal: '1000',
    name: 'Сбербанк руб',
    order: 3,
    price_usd: '0.014',
    reserve: '1000000',
    source: '1234 4321 1234 4321',
  },
]
