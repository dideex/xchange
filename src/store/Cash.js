import {observable, action, computed} from 'mobx'
import openSocket from 'socket.io-client'
import Api from '../components/Api'
import {noty} from '../components/common'

// Mobx Cash store
// Keeps the current transittion payment data
// And currency rate exchange

class Cash {
  @observable inputValue
  @observable outputValue
  @observable currencyInput
  @observable currencyOutput
  @observable paymentStatus
  @observable draggedBadgeCurrency
  @observable orderId
  @observable loading
  @observable isNetworkError
  @observable.ref currency

  constructor() {
    this.inputValue = 0
    this.outputValue = 0
    this.outputValueInUsd = 0
    this.currencyInput = 0
    this.currencyOutput = 2
    this.paymentStatus = 0 // 0 - null, 1 - created, 2 - sended, 3 - closed
    this.draggedBadgeCurrency = null
    this.lessThenMinimal = false
    this.moreThenReseved = false
    this.orderId = null
    this.loading = false
    this.currency = []
    this.isNetworkError = false
    this.userRate = 0.9
    this._setInitalData()
  }

  _setInitalData = () => {
    const url =
      process.env.NODE_ENV === 'development'
        ? 'http://localhost:3040'
        : 'http://176.119.158.145:3040'
    this.socket = openSocket.connect(url)
    this.errorEmitter = Api.errorEmitter.bind(this)
    this.fetchCurrency()
  }

  // finds an equal value for input value
  // for 1BTC (with rate 0.9) should be 0.9BTC
  _calcOutput = value =>
    (value * this.currency[this.currencyInput].price_usd * this.userRate) /
    this.currency[this.currencyOutput].price_usd

  // finds an equal value for output value
  // for 1BTC (with rate 0.9) should be 1BTC
  // cur bitc 5448.77817486
  // should be to 0.9 (usd 6054.198)
  _calcInput = value =>
    (value * this.currency[this.currencyOutput].price_usd) /
    (this.currency[this.currencyInput].price_usd * this.userRate)

  // Swaps currencies values when changed currency is already set
  _reverseInputs = () => {
    const temp = this.outputValue / this.userRate
    this.outputValue = this.inputValue * this.userRate
    this.inputValue = temp
  }

  // finds an equal value for output value
  // for 1BTC (wtih rate 0.9) should be 0.9BTC
  _calcOutputWithoutRates = value =>
    (value * this.currency[this.currencyInput].price_usd * this.userRate) /
    this.currency[this.currencyOutput].price_usd

  // finds an equal value for input value
  // for 1BTC (wtih rate 0.9) should be 0.9BTC
  _calcInputWithoutRates = value =>
    (value * this.currency[this.currencyOutput].price_usd) /
    (this.currency[this.currencyInput].price_usd * this.userRate)

  // It holds total amount in common unit for userprofile
  _calcOutputInUsd = () =>
    (this.outputValueInUsd =
      this.outputValue * this.currency[this.currencyOutput].price_usd)

  // check for a correct number
  _isNumber = num => /^\d+[.]?\d{0,3}$/.test(num)

  // It returns true when input's currencies are equals
  _isSameCurrencyLabel = (id1, id2) =>
    this.currency[id1].label === this.currency[id2].label

  // Parses nubmer for "0,2" to "0.2", and " 0,2  " to "0.2"
  _parseNumber = num => {
    if (num === '') return 0
    else {
      num = num.replace(/\s/g, '')
      num = num.replace(',', '.')
    }
    return num
  }

  // It recalculates both input to correct range
  @action('Verify values posible limit')
  correctValuesLimits = () => {
    const {minimal} = this.currency[this.currencyInput]
    const {reserve} = this.currency[this.currencyOutput]
    if (this.inputValue < +minimal) {
      this.inputValue = minimal
      this.outputValue = this._calcOutput(minimal)
    }
    if (this.outputValue * this.userRate > +reserve) {
      this.outputValue = reserve
      this.inputValue = this._calcInput(reserve)
    }
  }

  @action('clear message error field')
  clearErr = () => (this.errorMessage = null)

  // Change and validate input, and recalculate output value
  @action('change input')
  changeInput = (number = 0) => {
    this.clearErr()
    const parsedNumber = this._parseNumber(number)
    if (this._isNumber(parsedNumber)) {
      this.inputValue = parsedNumber
      this.outputValue = this._calcOutput(parsedNumber)
    }
    this._calcOutputInUsd()
  }

  // Change and validate output, and recalculate input value
  @action('change output')
  changeOutput = (number = 0) => {
    this.clearErr()
    const parsedNumber = this._parseNumber(number)
    if (this._isNumber(parsedNumber)) {
      this.outputValue = parsedNumber
      this.inputValue = this._calcInput(parsedNumber)
    }
    this._calcOutputInUsd()
  }

  // Change output value, or switch it with input currency when are the same
  @action('set currency output')
  setCurrencyOutput = (id = 0) => {
    this.clearErr()
    if (this.currencyOutput === +id) return null
    if (this._isSameCurrencyLabel(id, this.currencyInput)) {
      this.currencyInput = this.currencyOutput
      this.currencyOutput = +id
      this._reverseInputs()
    } else {
      this.currencyOutput = +id
      this.outputValue = this._calcOutputWithoutRates(this.inputValue)
    }
  }

  // Change output value, or switch it with input currency when are the same
  @action('set currency input')
  setCurrencyInput = (id = 0) => {
    this.clearErr()
    if (this.currencyInput === +id) return null
    if (this._isSameCurrencyLabel(id, this.currencyOutput)) {
      this.currencyOutput = this.currencyInput
      this.currencyInput = +id
      this._reverseInputs()
    } else {
      this.currencyInput = +id
      this.inputValue = this._calcInputWithoutRates(this.outputValue)
    }
  }

  /**
   * Creates payment
   * Refetch currency rate exchanges
   * When token is not undefiended, creates protected order, otherwise creates public
   * @param {token: {String}, fromWallet: {String}, toWallet: {String}, email: {String}}{Object}
   * @return <Promise>
   * @public
   */
  @action('create payment and push to server')
  createPayment = ({token, fromWallet, toWallet, email}) => {
    this.clearErr()
    this.isNetworkError = false
    this.loading = true
    this.correctValuesLimits()
    this._calcOutputInUsd()
    return this.fetchCurrency().then(async () => {
      this.paymentStatus = 1
      const data = {
        inputValue: this.inputValue,
        outputValue: this.outputValue,
        outputValueInUsd: this.outputValueInUsd,
        currencyInput: this.currency[this.currencyInput].name,
        currencyOutput: this.currency[this.currencyOutput].name,
        currencyInputLabel: this.currency[this.currencyInput].label,
        currencyOutputLabel: this.currency[this.currencyOutput].label,
        paymentStatus: 1,
        email,
        fromWallet,
        toWallet,
        token,
      }
      if (token)
        await Api.post('orders', data, token)
          .then(
            this.errorEmitter(({result}) => {
              this.orderId = result._id
              noty('Ваш перевод успешно создан')
            }),
          )
          .catch(err => {
            console.error(err)
            this.isNetworkError = true
            noty('Ошибка создания', 'error')
          })
      else
        await Api.post('guestOrders', data)
          .then(
            this.errorEmitter(({result}) => {
              this.orderId = result._id
              noty('Ваш перевод успешно создан')
            }),
          )
          .catch(err => {
            console.error(err)
            this.isNetworkError = true
            noty('Ошибка создания', 'error')
          })
      this.loading = false
    })
  }

  /**
   * Fetch currency data and currency rate exchange from the server
   * Sorting result currencies
   * Replace id to number instead a name
   * @return <Promise>
   * @public
   */
  @action('get currency from the server')
  fetchCurrency = () =>
    new Promise((resolve, reject) => {
      this.loading = true
      return Api.get('currency')
        .then(
          this.errorEmitter(({data, userRate}) => {
            this.userRate = userRate
            this.currency = Object.values(data)
              .sort((a, b) => a.order - b.order)
              .map((row, i) => ({...row, id: i}))
            this.loading = false
            resolve()
          }),
        )
        .catch(err => {
          noty('Ошибка сервера', 'error')
          console.error(err)
          this.loading = false
          reject()
        })
    })

  /**
   * Creates boadcast query to socketio
   * @param {emailinputValue{String} outputValue{Sting|Number} inputLabel{String|Number} outputLabel{String} currency{String} paymentStatus{Number}}<data>
   * @public
   */
  @action('emit socket')
  emitSocket = data => this.socket.emit('newOrder', data)

  /**
   * Creates POST query to the server which means change a payment status for orderd to 2(order transmitted)
   * @param email{Sting}
   * @public
   */
  @action('confirm payment')
  cofirmPayment = email => {
    this.paymentStatus = 2
    const data = {
      _id: this.orderId,
      currency: this.currency[this.currencyOutput].icon,
      value: this.outputValue,
      email,
    }
    Api.post('confirmOrder', data)
      .then(
        this.errorEmitter(() => {
          noty('Ваш перевод передан на обработку')
          this.emitSocket({
            email,
            currency: this.currency[this.currencyOutput].icon,
            inputValue: this.inputValue,
            outputValue: this.outputValue,
            inputLabel: this.currency[this.currencyInput].label,
            outputLabel: this.currency[this.currencyOutput].label,
            paymentStatus: this.paymentStatus,
          })
        }),
      )
      .catch(err => {
        console.error(err)
        noty('Ошибка перевода', 'error')
      })
  }

  // helper for react-dnd, memoize the id which was dragged
  @action('drag badge')
  handleDragBadge = id => (this.draggedBadgeCurrency = id)

  // return formatted string with input value and their currency label is
  @computed
  get getInput() {
    return `${this.inputValue}, ${this.currency[this.currencyInput].label.toUpperCase()}`
  }
  // return formatted string with output value and their currency label is
  @computed
  get getOutput() {
    return `${this.outputValue}, ${this.currency[
      this.currencyOutput
    ].label.toUpperCase()}`
  }
  //returns formatted minimal amount for input value is
  @computed
  get getMinimalAmount() {
    const formatter = new Intl.NumberFormat('ru', 'currency')
    if (!this.currency.length) return null
    return `${formatter.format(this.currency[this.currencyInput].minimal)} ${
      this.currency[this.currencyInput].label
    }`
  }
  //returns formatted reserved amount for output value is
  @computed
  get getCurrencyReserve() {
    const formatter = new Intl.NumberFormat('ru', 'currency')
    if (!this.currency.length) return null
    return `${formatter.format(this.currency[this.currencyOutput].reserve)} ${
      this.currency[this.currencyOutput].label
    }`
  }
}

export const cash = new Cash()

export default Cash
