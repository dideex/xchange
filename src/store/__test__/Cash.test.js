import Cash from '../Cash'
import Api from '../../components/Api'
import openSocket from 'socket.io-client'
import currencies from '../__mocks__/currencies.json'
import {fakeToken, fakeUser} from '../__mocks__/data'
// jest.mock('socket.io-client', jest.fn)

jest.mock('../../components/Api', () => ({
  post: () => {},
  get: () => Promise.resolve({data: {}}),
  errorEmitter: jest.fn(data => fn => data(fn)),
}))

const sortedData = [
  {order: 1, id: 0, name: 'eth'},
  {order: 2, id: 1, name: 'bitcoin'},
  {order: 3, id: 2, name: 'rur'},
]

const unsortedData = [
  {order: 2, _id: 2, id: 'bitcoin', name: 'lisk'},
  {order: 1, _id: 1, id: 'eth', name: 'tether'},
  {order: 3, _id: 3, id: 'rur', name: 'litecoin'},
]

const fakeData = {
  data: sortedData,
  userRate: 0.9,
}
const fakeUnsortData = {
  data: unsortedData,
  userRate: 0.9,
}

describe('Cash store tests', () => {
  let store
  beforeEach(() => {
    store = new Cash()
  })

  it('jest', () => {
    expect(1 + 1).toBe(2)
  })

  it('Fetch currency data', async () => {
    Api.get = jest.fn(() => Promise.resolve(fakeData))

    await store.fetchCurrency()
    expect(Api.get).toHaveBeenCalledTimes(1)
    expect(Api.get).toHaveBeenCalledWith('currency')
    expect(store.userRate).toBe(fakeData.userRate)
    expect(store.currency).toEqual(expect.arrayContaining(fakeData.data))
    expect(store.loading).toBe(false)
  })

  it('Fetch currency data and sort it', async () => {
    Api.get = jest.fn(() => Promise.resolve(fakeUnsortData))

    await store.fetchCurrency()
    const sortedData = Object.values(fakeUnsortData.data)
      .sort((a, b) => a.order - b.order)
      .map((row, i) => ({...row, id: i}))
    expect(store.currency).toEqual(expect.arrayContaining(sortedData))
  })

  it('Post new payment with token', async () => {
    Api.post = jest.fn(() => Promise.resolve({result: unsortedData[0]}))
    store.fetchCurrency = jest.fn(() => Promise.resolve({data: {}}))
    store.correctValuesLimits = jest.fn()
    store._calcOutputInUsd = jest.fn()
    store.clearErr = jest.fn()

    const fakeData = {
      inputValue: store.inputValue,
      outputValue: store.outputValue,
      outputValueInUsd: store.outputValueInUsd,
      currencyInput: store.currency[store.currencyInput].name,
      currencyOutput: store.currency[store.currencyOutput].name,
      currencyInputLabel: store.currency[store.currencyInput].label,
      currencyOutputLabel: store.currency[store.currencyOutput].label,
      paymentStatus: 1,
      email: fakeUser.username,
      fromWallet: '1234',
      toWallet: '4321',
      token: fakeToken,
    }

    await store.createPayment({
      token: fakeToken,
      fromWallet: '1234',
      toWallet: '4321',
      email: fakeUser.username,
    })
    expect(store.fetchCurrency).toHaveBeenCalledTimes(1)
    expect(store.correctValuesLimits).toHaveBeenCalledTimes(1)
    expect(store._calcOutputInUsd).toHaveBeenCalledTimes(1)
    expect(store.clearErr).toHaveBeenCalledTimes(1)
    expect(Api.post).toHaveBeenCalledTimes(1)
    expect(Api.post).toHaveBeenCalledWith('orders', fakeData, fakeToken)
    expect(store.orderId).toBe(unsortedData[0]._id)
    expect(store.paymentStatus).toBe(1)
  })

  it('Post new payment without token', async () => {
    Api.post = jest.fn(() => Promise.resolve({result: unsortedData[0]}))
    store.fetchCurrency = jest.fn(() => Promise.resolve({data: {}}))
    store.correctValuesLimits = jest.fn()
    store._calcOutputInUsd = jest.fn()
    store.clearErr = jest.fn()

    const fakeData = {
      inputValue: store.inputValue,
      outputValue: store.outputValue,
      outputValueInUsd: store.outputValueInUsd,
      currencyInput: store.currency[store.currencyInput].name,
      currencyOutput: store.currency[store.currencyOutput].name,
      currencyInputLabel: store.currency[store.currencyInput].label,
      currencyOutputLabel: store.currency[store.currencyOutput].label,
      paymentStatus: 1,
      email: fakeUser.username,
      fromWallet: '1234',
      toWallet: '4321',
      token: null,
    }

    await store.createPayment({
      token: null,
      fromWallet: '1234',
      toWallet: '4321',
      email: fakeUser.username,
    })
    expect(Api.post).toHaveBeenCalledTimes(1)
    expect(Api.post).toHaveBeenCalledWith('guestOrders', fakeData)
    expect(store.orderId).toBe(unsortedData[0]._id)
  })

  it('Confirm payment', async() => {
    Api.post = jest.fn(() => Promise.resolve())
    store.emitSocket = jest.fn()

    store.orderId = unsortedData[0]._id
    const fakeData = {
      _id: unsortedData[0]._id,
    }

    await store.cofirmPayment(fakeUser.username)
  })
})

describe('Cash store currencies behaviour', () => {
  let store

  beforeEach(() => {
    Api.get = jest.fn(() => Promise.resolve(currencies))
    store = new Cash()
  })

  it('Init currency', () => {
    expect(store.userRate).toBe(currencies.userRate)
    expect(store.currencyInput).toBe(0)
    expect(store.currencyOutput).toBe(2)
    store.currency.forEach((currency, i) => {
      expect(currency).toMatchObject(currencies.data[i])
    })
  })

  it('Swap currencies', () => {
    store.setCurrencyOutput(0)
    expect(store.currencyInput).toBe(2)
    expect(store.currencyOutput).toBe(0)
    store.setCurrencyOutput(2)
    expect(store.currencyInput).toBe(0)
    expect(store.currencyOutput).toBe(2)
    store.setCurrencyOutput(1)
    expect(store.currencyInput).toBe(0)
    expect(store.currencyOutput).toBe(1)
    store.setCurrencyOutput(0)
    expect(store.currencyInput).toBe(1)
    expect(store.currencyOutput).toBe(0)
  })

  it('Sets correct wallets for currency', () => {
    store.setCurrencyInput(0)
    expect(store.currency[store.currencyInput].source).toBe(currencies.data[0].source)
    store.setCurrencyInput(1)
    expect(store.currency[store.currencyInput].source).toBe(currencies.data[1].source)
    store.setCurrencyInput(2)
    expect(store.currency[store.currencyInput].source).toBe(currencies.data[2].source)
  })
})

describe('UI elements behaviour', () => {
  let store
  beforeEach(() => {
    Api.get = jest.fn(() => Promise.resolve(currencies))
    store = new Cash()
  })

  it('Change inputs', () => {
    store.changeInput('100')
    expect(store.inputValue).toBe('100')
    store.changeInput('200')
    expect(store.inputValue).toBe('200')
    store.changeOutput('100')
    expect(store.outputValue).toBe('100')
    store.changeOutput('200')
    expect(store.outputValue).toBe('200')
  })

  it('Mutation float numbers', () => {
    const base = '100'
    store.changeInput(base)
    store.changeInput('100q')
    expect(store.inputValue).toBe(base)
    store.changeInput(base + '.')
    expect(store.inputValue).toBe(base + '.')
    store.changeInput(base + '..')
    expect(store.inputValue).toBe(base + '.')
    store.changeInput(base + ',')
    expect(store.inputValue).toBe(base + '.')
    store.changeInput(base + ',0')
    expect(store.inputValue).toBe(base + '.0')
    store.changeInput(base + ',,0')
    expect(store.inputValue).toBe(base + '.0')
  })
})

describe('Currnecy calculate behaviour', () => {
  let store
  const btcRate = currencies.data[0].price_usd * currencies.userRate
  const ethRate = currencies.data[1].price_usd * currencies.userRate
  const rurRate = currencies.data[2].price_usd * currencies.userRate

  const btcPrice = currencies.data[0].price_usd
  const ethPrice = currencies.data[1].price_usd
  const rurPrice = currencies.data[2].price_usd

  beforeEach(() => {
    Api.get = jest.fn(() => Promise.resolve(currencies))
    store = new Cash()
  })

  it('Change input value', () => {
    store.changeInput('100')
    expect(store.inputValue).toBe('100')
    expect(store.outputValue).toBe(store.inputValue * btcRate)
    store.changeInput('200')
    expect(store.inputValue).toBe('200')
    expect(store.outputValue).toBe(store.inputValue * btcRate)
    store.changeInput('800')
    expect(store.inputValue).toBe('800')
    expect(store.outputValue).toBe(store.inputValue * btcRate)
  })

  it('Change output value', () => {
    store.changeOutput('100')
    expect(store.outputValue).toBe('100')
    expect(store.inputValue).toBe(store.outputValue / (btcPrice * currencies.userRate))
    store.changeOutput('400')
    expect(store.outputValue).toBe('400')
    expect(store.inputValue).toBe(store.outputValue / (btcPrice * currencies.userRate))
    store.changeOutput('500')
    expect(store.outputValue).toBe('500')
    expect(store.inputValue).toBe(store.outputValue / (btcPrice * currencies.userRate))
  })

  it('Change input currency', () => {
    store.changeOutput('100')
    expect(store.outputValue).toBe('100')
    expect(store.inputValue).toBe(store.outputValue / (btcPrice * currencies.userRate))
    store.setCurrencyInput(1)
    expect(store.inputValue).toBe(store.outputValue / (ethPrice * currencies.userRate))
    store.changeOutput('200')
    expect(store.inputValue).toBe(store.outputValue / (ethPrice * currencies.userRate))
    store.setCurrencyInput(0)
    expect(store.inputValue).toBe(store.outputValue / (btcPrice * currencies.userRate))
  })

  it('Change output currency', () => {
    store.changeInput('100')
    store.setCurrencyOutput(1)
    expect(store.inputValue).toBe('100')
    expect(store.outputValue).toBe(store.inputValue * ethRate)
    store.setCurrencyOutput(2)
    expect(store.outputValue).toBe(store.inputValue * btcRate)
    store.changeInput('400')
    store.setCurrencyOutput(1)
    expect(store.inputValue).toBe('400')
    expect(store.outputValue).toBe(store.inputValue * ethRate)
    store.setCurrencyOutput(2)
    expect(store.outputValue).toBe(store.inputValue * btcRate)
  })

  it('Swap input to output', () => {
    store.changeInput('100')
    store.setCurrencyInput(2)
    expect(store.inputValue).toBe(100 * btcPrice)
    expect(store.outputValue).toBe(100 * store.userRate)
    store.changeInput('100')
    store.setCurrencyInput(0)
    expect(store.inputValue).toBe(100 / btcPrice)
    expect(store.outputValue).toBe(100 * store.userRate)
    store.setCurrencyOutput(1)
    store.changeInput('100')
    store.setCurrencyInput(1)
    expect(store.inputValue).toBe(100 * ethPrice)
    expect(store.outputValue).toBe(100 * store.userRate)
    store.changeInput('100')
    store.setCurrencyInput(0)
    expect(store.inputValue).toBe(100 / ethPrice)
    expect(store.outputValue).toBe(100 * store.userRate)
  })

  it('Swap output to input', () => {
    store.changeInput('100')
    store.setCurrencyOutput(0)
    expect(store.inputValue).toBe(100 * btcPrice)
    expect(store.outputValue).toBe(100 * store.userRate)
    store.changeInput('100')
    store.setCurrencyOutput(2)
    expect(store.inputValue).toBe(100 / btcPrice)
    expect(store.outputValue).toBe(100 * store.userRate)
    store.setCurrencyInput(1)
    store.changeInput('100')
    store.setCurrencyOutput(1)
    expect(store.inputValue).toBe(100 * ethPrice)
    expect(store.outputValue).toBe(100 * store.userRate)
    store.changeInput('100')
    store.setCurrencyOutput(2)
    expect(store.inputValue).toBe(100 / ethPrice)
    expect(store.outputValue).toBe(100 * store.userRate)
  })

  it('Set value less then minimal payment', () => {
    store.setCurrencyInput(2)
    store.changeInput('100')
    store.correctValuesLimits()
    expect(store.inputValue).toBe('1000')
    store.setCurrencyInput(1)
    store.changeInput('0.5')
    store.correctValuesLimits()
    expect(store.inputValue).toBe('1')
  })

  it('Set value bigger then reserved payment', () => {
    store.changeOutput('9999999')
    store.correctValuesLimits()
    expect(store.outputValue).toBe('1000000')
    store.setCurrencyOutput(1)
    store.changeOutput('9999999')
    store.correctValuesLimits()
    expect(store.outputValue).toBe('10000')
  })
})
