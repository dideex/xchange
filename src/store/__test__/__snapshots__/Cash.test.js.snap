// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currnecy calculate behaviour Output value in usd behaviour 1`] = `720`;

exports[`Currnecy calculate behaviour Output value in usd behaviour 2`] = `1440`;

exports[`Currnecy calculate behaviour Output value in usd behaviour 3`] = `200`;

exports[`Currnecy calculate behaviour Output value in usd behaviour 4`] = `400`;

exports[`UI elements behaviour Format minimal amount 1`] = `"0.1 BTC"`;

exports[`UI elements behaviour Format minimal amount 2`] = `"1 ETH"`;

exports[`UI elements behaviour Format minimal amount 3`] = `"1,000 RUR"`;

exports[`UI elements behaviour Format reserved amount 1`] = `"1,000,000 RUR"`;

exports[`UI elements behaviour Format reserved amount 2`] = `"10,000 BTC"`;

exports[`UI elements behaviour Format reserved amount 3`] = `"10,000 ETH"`;
